File: archive/2024-12-06/tree/T14_findBottomLeftValue.java
package org.review2024winter.tree;

public class T14_findBottomLeftValue {
    int maxDepth = 0;
    int res = 0;

    public int findBottomLeftValue(TreeNode root) {
        traversal(root, 1);
        return res;
    }

    public void traversal(TreeNode root, int depth) {
        if (root == null)
            return;
        if (root.left == null && root.right == null && depth > maxDepth) {
            maxDepth = depth;
            res = root.val;
            return;
        }
        traversal(root.left, depth + 1);
        traversal(root.right, depth + 1);
    }
}

File: archive/2026-04-05/ch06_tree/t14_find_bottom_left_value.py
"""LeetCode 513. Find Bottom Left Tree Value
https://leetcode.cn/problems/find-bottom-left-tree-value/
"""

from typing import Optional, Deque

from cytoolz.itertoolz import deque

from tree_node import TreeNode


class Solution:
    def findBottomLeftValue(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        ans: int = 0
        queue: Deque[TreeNode] = deque([root])

        while queue:
            ans = queue[0].val
            for _ in range(len(queue)):
                curr = queue.popleft()
                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)

        return ans


if __name__ == "__main__":
    solution = Solution()
    root1 = TreeNode(2, TreeNode(1), TreeNode(3))
    assert solution.findBottomLeftValue(root1) == 1

    root2 = TreeNode(
        1, TreeNode(2, TreeNode(4)), TreeNode(3, TreeNode(5, TreeNode(7)), TreeNode(6))
    )
    assert solution.findBottomLeftValue(root2) == 7

    print("All tests passed.")
